File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        int amount = lists.length;
        int interval = 1;

        while( interval < amount ){
            for(int i=0; i<amount-interval; i+=interval*2){
                lists[i]=merge2Lists(lists[i],lists[i+interval]);
            }
            interval *= 2;
        }
        return amount > 0 ? lists[0] : null;
    }

    public ListNode merge2Lists(ListNode list1, ListNode list2){
        ListNode dummy = new ListNode(-1);
        ListNode tail = dummy;

        while(list1!= null && list2!= null){
            if(list1.val <= list2.val){
                tail.next =list1;
                list1=list1.next;
            }
            else{
                tail.next =list2;
                list2=list2.next;
            }
            tail =tail.next;
        }

        if(list1!=null){
            tail.next =list1;
        }
        if(list2!=null){
            tail.next =list2;
        }

        return dummy.next;
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []

        for i , node in enumerate(lists):
            if node:
                heapq.heappush(heap,(node.val,i,node))

        dummy = ListNode(0)
        current = dummy
        
        # 2. PROCESS HEAP
        while heap:
            val, i, node = heapq.heappop(heap)
            
            # Link the node to our result
            current.next = node
            current = current.next
            
            # Push the NEXT node from the same list into heap
            if node.next:
                heapq.heappush(heap, (node.next.val, i, node.next))
                
        return dummy.next